src/common/commonutils/CommandUtils.c

```diff
@@ -3,296 +3,235 @@
 
 #include "Internal.h"
 
-static void KillProcess(pid_t processId)
+#include <sys/select.h>
+
+static long MonotonicTime()
 {
-    fflush(NULL);
-    if (processId > 0)
+    struct timespec ts;
+    if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0)
     {
-        kill(processId, SIGKILL);
+        return (long)(ts.tv_sec);
     }
+    return -1;
 }
 
-static int NormalizeStatus(int status)
+#define BUFFER_SIZE 1024
+
+int ExecuteCommand(void* context, const char* command, bool replaceEol, bool forJson, unsigned int maxTextResultBytes, unsigned int timeoutSeconds,
+    char** textResult, CommandCallback callback, OsConfigLogHandle log)
 {
-    int newStatus = status;
-    if ((ETIME != newStatus) && (ECANCELED != newStatus))
+    int workerPid = -1;
+    int pipefd[2] = {0};
+    const int defaultCommandTimeout = 60;  // seconds
+    const int callbackIntervalSeconds = 5; // seconds
+    const struct timeval selectInterval = {0, 100 * 1000}; // 100 ms, accuracy of timeouts.
+    struct timeval tv;
+    int status = -1;
+    int childStatus = 0;
+    int bytesRead = 0;
+    int inputBufferPos = 0;
+    unsigned int outputBufferPos = 0;
+    unsigned int outputBufferSize = 0;
+    long startTime = 0;
+    long lastCallbackTime = 0;
+    long currentTime = 0;
+    char buffer[BUFFER_SIZE];
+    fd_set fdset;
+    int ret = 0;
+    char* tmp = NULL;
+
+
+    if (NULL == command)
     {
-        if (WIFEXITED(newStatus))
-        {
-            newStatus = WEXITSTATUS(newStatus);
-        }
-        else
-        {
-            newStatus = errno ? errno : -1;
-        }
+        OsConfigLogDebug(log, "Command cannot be NULL");
+        return -1;
     }
-    return newStatus;
-}
-
-static int SystemCommand(void* context, const char* command, int timeoutSeconds, CommandCallback callback, OsConfigLogHandle log)
-{
-    const int callbackIntervalSeconds = 5; //seconds
-    const int defaultCommandTimeout = 60; //seconds
+    // Create a pipe, then fork. Forked process duplicates the write pipe end to stdout and stderr,
+    // then execs the shell with the given command. The main process uses select() with a timeout.
```

Comment:
>with a timeout [](http://example.com/codeflow?start=83&length=14)

Speaking of 'a timeout', this is not any timeout, but the one specified by the function 'timeoutSeconds' argument, correct?

Reply:
In this case that's the selectInterval.